### PR TITLE
Heretic: implement new game

### PR DESCRIPTION
**Pull request by Daivuk**
Title3

## What is this fixing or adding?

Self explanatory title.

Note to reviewers: All source files except for __init__.py are generated through a tool. They will be formatted the same way as DOOM 1993. I wouldn't waste time reviewing these as the format was already approved. The __init__.py is almost similar to DOOM 1993 except for minor changes. Because Heretic is also Episodic, this code has more similarities with DOOM 1993 than DOOM II.

The APPDOOM client can be found here: https://github.com/Daivuk/apdoom/releases

## How was this tested?

Community tested for months. But this code adds new stuff, and still need to be tested a bit longer. The apworld was given to the community. This commit also adds 2 new episodes: Episode 4 and 5.

## If this makes graphical changes, please attach screenshots.

The AP icons used in game are the same as DOOM 1993. But it is stored in a different wad file APHERETIC.WAD. They have different pallet.
app5
ap4

The HUB teleporter is also different than DOOM 1993 and DOOM II to fit the Heretic theme:
LVST2

Level select is the same principal as in DOOM 1993:
image

### CODE REVIEW

File: worlds/heretic/__init__.py [hunk 1]
```diff
@@ -0,0 +1,287 @@
...
+        "Enchanted Shield": 5,
+        "Morph Ovum": 3,
+        "Mystic Urn": 2,
+        "Chaos Device": 1,
+        "Ring of Invincibility": 1,
+        "Shadowsphere": 1
+    }
+
+    def __init__(self, world: MultiWorld, player: int):
+        self.included_episodes = [1, 1, 1, 0, 0] # Temp, only first episode is coded
```
Comment (Berserker66) [inline worlds/heretic/__init__.py:81]: this comment looks outdated

**Comment by ThePhar** [comment]
Can you add this world to `docs/CODEOWNERS` and `README.md`?

File: worlds/heretic/__init__.py [hunk 2]
```diff
@@ -0,0 +1,287 @@
...
+
+    def create_ratioed_items(self, item_name: str, itempool: List[HereticItem]):
+        remaining_loc = self.location_count - len(itempool)
+        if remaining_loc <= 0:
+            return
+
+        episode_count = self.get_episode_count()
+        count = min(remaining_loc, max(1, self.items_ratio[item_name] * episode_count))
+        if count == 0:
+            logger.warning("Warning, no ", item_name, " will be placed.")
```
Comment (Berserker66) [inline worlds/heretic/__init__.py:271]: logging does not concatenate string arguments like print, this will crash.

**Comment by Daivuk** [comment]
> Can you add this world to `docs/CODEOWNERS` and `README.md`?

done

File: worlds/heretic/__init__.py [hunk 3]
```diff
@@ -0,0 +1,287 @@
...
+        count = min(remaining_loc, max(1, self.items_ratio[item_name] * episode_count))
+        if count == 0:
+            logger.warning("Warning, no " + item_name + " will be placed.")
+            return
+
+        for i in range(count):
+            itempool.append(self.create_item(item_name))
+
+    def fill_slot_data(self) -> Dict[str, Any]:
+        slot_data = {name: getattr(self.multiworld, name)[self.player].value for name in self.option_definitions}
```
Comment (Berserker66) [inline worlds/heretic/__init__.py:278]: Do you really need all option results, including things like allow death logic?

Comment (Daivuk) [reply worlds/heretic/__init__.py:278]: No, I'll do like I had to do for DOOM II:
```python
        return self.options.as_dict("difficulty", "random_monsters", "random_pickups", "random_music", "flip_levels", "allow_death_logic", "pro", "death_link", "reset_level_on_death", "episode1", "episode2", "episode3", "episode4")```

**Review by Berserker66 — APPROVED** [review]

**Pull request merged**